graph/src/components/metrics/aggregate.rs

```diff
@@ -0,0 +1,96 @@
+use std::collections::HashMap;
+use std::time::Duration;
+
+use crate::prelude::*;
+
+/// Aggregate metrics for a single value.
+///
+/// Given a `name`, this registers the following gauges:
+///
+/// - `{name}_count` (counting all updates to the value)
+/// - `{name}_sum` (the sum of all updates to the value over time)
+/// - `{name}_avg` (the moving average of all updates to the value over time)
+/// - `{name}_cur` (the current or latest update to the value)
+pub struct Aggregate {
+    /// Number of values.
+    n: u64,
+
+    /// Sum over all values.
+    sum: f64,
+
+    /// Moving average over the values.
+    avg: f64,
+
+    /// Latest value.
+    cur: f64,
+
+    /// Metrics for Prometheus.
+    count_gauge: Box<Gauge>,
+    sum_gauge: Box<Gauge>,
+    avg_gauge: Box<Gauge>,
+    cur_gauge: Box<Gauge>,
+}
+
+impl Aggregate {
+    pub fn new(name: String, help: &str, registry: Arc<dyn MetricsRegistry>) -> Self {
+        let count_gauge = registry
+            .new_gauge(
+                format!("{}_count", name),
+                format!("{} (count)", help),
+                HashMap::new(),
+            )
+            .expect(format!("failed to register metric `{}_count`", name).as_str());
+
+        let sum_gauge = registry
+            .new_gauge(
+                format!("{}_sum", name),
+                format!("{} (sum)", help),
+                HashMap::new(),
+            )
+            .expect(format!("failed to register metric `{}_sum`", name).as_str());
+
+        let avg_gauge = registry
+            .new_gauge(
+                format!("{}_avg", name),
+                format!("{} (avg)", help),
+                HashMap::new(),
+            )
+            .expect(format!("failed to register metric `{}_avg`", name).as_str());
+        let cur_gauge = registry
+            .new_gauge(
+                format!("{}_cur", name),
+                format!("{} (cur)", help),
+                HashMap::new(),
+            )
+            .expect(format!("failed to register metric `{}_cur`", name).as_str());
```

Comment:
We do this in other places as well, but it makes me nervous that the graph node would crash because of a metrics bug.